### PR TITLE
APEXMALHAR-2008: Create HDFS File Reader module

### PR DESCRIPTION
**Pull request by DT-Priyanka**
Code to add HDFS file reader module. 
1. The module reads file/list of files (directory is also accepted) and emit the file blocks. 
2. The module can be configured to emit blocks in order or out of order.
3. Module reads file blocks in parallel. The number of parallel readers is configurable, if not configured it will increase or decrease readers dynamically as per input data rate.

Also updated code of FileSplitterInput to add some improvements:
1. Tracking last file reference times of each folder differently, to avoid duplicates (duplicates could be due to same relative paths of multiple files/sub dir)
2. Small improvements in code.

### CODE REVIEW

**Comment by sandeepdeshmukh** [comment]
All the white space related changes should reverted. There are number of them.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /**
+   * Converts Scheme part of the URI to lower case. Multiple URI can be comma separated. If no scheme is there, no
+   * change is made.
+   * 
+   * @param
+   * @return String with scheme part as lower case
+   */
+  private static String convertSchemeToLowerCase(String uri)
```
Comment (sandeepdeshmukh) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Is there a test case covering  this?

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * under the License.
+ */
+package com.datatorrent.lib.io.block;
+
+/**
+ * HDFSBlockMetadata extends {@link BlockMetadata} to provide an option if blocks of a file should be read in-sequence
+ * or in-parallel
+ */
+public class HDFSBlockMetadata extends BlockMetadata.FileBlockMetadata
+{
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java]: Is this class is specific to only HDFS? If not, change the class name.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  public boolean isReadBlockInSequence()
+  {
+    return readBlockInSequence;
+  }
+
+  public void setReadBlockInSequence(boolean readBlockInSequence)
+  {
+    this.readBlockInSequence = readBlockInSequence;
+  }
+
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java]: Add java docs for setters and getters.

**Comment by sandeepdeshmukh** [comment]
Java doc is missing at many places.

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java [hunk 4]
```diff
@@ -384,7 +399,14 @@ protected void scanIterationComplete()
       lastScanMillis = System.currentTimeMillis();
     }
 
-    protected void scan(@NotNull Path filePath, Path rootPath)
+    protected void scan(Path filePath, Path rootPath)
```
Comment (sandeepdeshmukh) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java]: Why is NotNull removed?

File: library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java [hunk 5]
```diff
@@ -215,7 +213,7 @@ public void testTimeScan() throws InterruptedException, IOException, TimeoutExce
     testMeta.blockMetadataSink.clear();
 
     Thread.sleep(1000);
-    //added a new relativeFilePath
+    // added a new relativeFilePath
```
Comment (sandeepdeshmukh) [inline library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java]: Example of white space change. Please revert this and all such changes.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+/**
+ * HDFSInputModule is used to read files from HDFS. <br/>
+ * Module emits FileMetadata, BlockMetadata and the block bytes.
+ */
+public class HDFSInputModule implements Module
+{
+
+  @NotNull
+  @Size(min = 1)
+  private String files;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: This is specific to HDFS. So, Split the files property to host, port and files/directories to copy.

Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: For splitting property to host, port, directory there should be discussion on dev@apex. Else, stick to original convention. We would split it later only if required.

Comment (sandeepdeshmukh) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: The purpose of Modules is to ease configuration of operators and hence it is good to separate host, port, directory etc.

Comment (yogidevendra) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: @sandeepdeshmukh If we decide to do this; this should be discussed explicitly on discussion forum. Because, the discussion conclusion may impact many other modules. This point is not specific to this PR and should be conveyed to wider audience.

**Comment by chinmaykolhatkar** [comment]
Can you please check why travis build is failing?

Failed tests: 
  HDFSInputModuleAppTest.testApplication:126->verifyFileContents:140 File data doesn't contain expected text

**Comment by chinmaykolhatkar** [comment]
I suggest all the new files should be moved under org.apache.apex namespace.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+package com.datatorrent.lib.io.block;
+
+import java.io.IOException;
+import java.net.URI;
+
+import org.apache.hadoop.fs.FileSystem;
+
+import com.google.common.base.Splitter;
+
+public class HDFSBlockReader extends FSSliceReader
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Javadoc for class is missing. Please add it.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  public String getUri()
+  {
+    return uri;
+  }
+
+  /**
+   * Converts Scheme part of the URI to lower case. Multiple URI can be comma separated. If no scheme is there, no
+   * change is made.
+   * 
+   * @param
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Please give description for param uri in javadoc.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java (hunk 1 above)
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: What is the reason for this method to be static? If this is a utility type method can this be moved to somewhere else more accessible?

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+   * @return String with scheme part as lower case
+   */
+  private static String convertSchemeToLowerCase(String uri)
+  {
+    if (uri == null) {
+      return null;
+    }
+    StringBuilder inputMod = new StringBuilder();
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: A better name than "inputMod"?

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    for (String f : Splitter.on(",").omitEmptyStrings().split(uri)) {
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Considering this is a utility method, I suggest not to take a comma seperated URIs here. What if someone has some other delimiter? 
I think from design perspective its better for this method to take a single URI and let the caller of this method take care of seperator.

File: library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java (hunk 4 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/FileSplitterInput.java]: By Mistake, will revert this.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+import javax.validation.constraints.NotNull;
+
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.lib.io.block.BlockMetadata.FileBlockMetadata;
+import com.datatorrent.lib.io.block.HDFSBlockMetadata;
+
+public class HDFSFileSplitter extends FileSplitterInput
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:35]: Class level javadoc is missing. Not having this causes problem in release process. Please add it.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+  {
+    return sequencialFileRead;
+  }
+
+  public void setSequencialFileRead(boolean sequencialFileRead)
+  {
+    this.sequencialFileRead = sequencialFileRead;
+  }
+
+  public static class HDFSScanner extends TimeBasedDirectoryScanner
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:68]: Can you please provide class level javadoc?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    {
+      return new Path(filePathStr).toUri().getPath().contains(UNSUPPORTED_CHARACTOR);
+    }
+  }
+
+  public static class HDFSFileMetaData extends FileMetadata
+  {
+    private String relativePath;
+
+    protected HDFSFileMetaData()
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: Any reason why empty constructor is protected but other constructor is private?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+    public void setRelativePath(String relativePath)
+    {
+      this.relativePath = relativePath;
+    }
+
+    @Override
+    public String toString()
+    {
+      return "HDFSFileMetaData [relativePath=" + relativePath + ", getNumberOfBlocks()=" + getNumberOfBlocks() + ", getFileName()=" + getFileName() + ", getFileLength()=" + getFileLength() + ", isDirectory()=" + isDirectory() + "]";
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: Please have variable names in toString and not method names.
For example change getNumberOfBlocks with numberOfBlocks etc..

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+import com.datatorrent.lib.io.block.HDFSBlockReader;
+import com.datatorrent.lib.io.fs.AbstractFileSplitter.FileMetadata;
+import com.datatorrent.lib.io.fs.HDFSFileSplitter.HDFSScanner;
+import com.datatorrent.netlet.util.Slice;
+
+/**
+ * HDFSInputModule is used to read files from HDFS. <br/>
+ * Module emits FileMetadata, BlockMetadata and the block bytes.
+ */
+public class HDFSInputModule implements Module
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:51]: Can you please describe what you've mentioned in this PR description as a javadoc in HDFSInputModule?
Its good to mention in default what all this module is responsible for and also how it'll work.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    dag.setAttribute(blockReader, Context.OperatorContext.METRICS_AGGREGATOR, blockReaderMetrics);
+    dag.setAttribute(blockReader, Context.OperatorContext.COUNTERS_AGGREGATOR, new BasicCounters.LongAggregator<MutableLong>());
+  }
+
+  /**
+   * A comma separated list of directories to scan. If the path is not fully qualified the default file system is used.
+   * A fully qualified path can be provided to scan directories in other filesystems.
+   *
+   * @param files
+   *          files
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:107]: Can we please have javadoc comments in proper format?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+{
+
+  @NotNull
+  @Size(min = 1)
+  private String files;
+  private String filePatternRegularExp;
+  @Min(0)
+  private long scanIntervalMillis;
+  private boolean recursive = true;
+  private long blockSize;
```
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:61]: Any default for the blockSize?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 15 above)
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:51]: As a part of this comment, can you provide what are the mandatory properties for this modules and what are optional?
OR atleast in the javadoc for respective setter method.

File: library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java [hunk 18]
```diff
@@ -110,8 +110,7 @@ protected void starting(org.junit.runner.Description description)
       fileSplitterInput.setIdempotentStorageManager(new IdempotentStorageManager.NoopIdempotentStorageManager());
 
       Attribute.AttributeMap.DefaultAttributeMap attributes = new Attribute.AttributeMap.DefaultAttributeMap();
-      attributes.put(Context.DAGContext.APPLICATION_PATH,
-          "target/" + className + "/" + methodName + "/" + Long.toHexString(System.currentTimeMillis()));
+      attributes.put(Context.DAGContext.APPLICATION_PATH, "target/" + className + "/" + methodName + "/" + Long.toHexString(System.currentTimeMillis()));
```
Comment (chinmaykolhatkar) [inline library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java]: Please revert all the changes in this file which are not related to this PR. Only take care of Checkstyle violations which appears in new code.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java (hunk 10 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Won't keep it as utility method as no one else is going to use if as of now.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 13 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: super class is doing that, I am just carrying over the same thing.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 14 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: This is auto generated code, would prefer not to change it.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 6 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: Won't split the properties for now, this could create problems if someone integrates operators/modules with external UI.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 17 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:61]: Default will be set by operator based on HDFS block size.

File: library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java (hunk 18 above)
Comment (DT-Priyanka) [inline library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java]: There are no checkstyle violations in new code, still the violation count was more than limit, so fixed some violations in this file.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java (hunk 2 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockMetadata.java]: removing the class and moving property to parent class.

File: library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java (hunk 1 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/HDFSBlockReader.java]: Adding unit test.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 14 above)
Comment (chinmaykolhatkar) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: auto generated by what?

File: library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java (hunk 18 above)
Comment (chinmaykolhatkar) [inline library/src/test/java/com/datatorrent/lib/io/fs/FileSplitterInputTest.java]: That's very wierd considering no code gets into master without rat check AFAIK.
 So without these changes does checkstyle validation fails?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+  private boolean sequencialFileRead = false;
+  private int readersCount;
+
+  public final transient ProxyOutputPort<FileMetadata> filesMetadataOutput = new ProxyOutputPort();
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java]: Please use <> for generic type in instantiation: new ProxyOutputPort<>()

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+
+    blockReader.setUri(files);
+    if (readersCount != 0) {
+      dag.setAttribute(blockReader, Context.OperatorContext.PARTITIONER, new StatelessPartitioner<BlockReader>(readersCount));
+    }
+
+    MetricsAggregator blockReaderMetrics = new MetricsAggregator();
+    blockReaderMetrics.addAggregators("bytesReadPerSec", new SingleMetricAggregator[] {new LongSumAggregator()});
+    dag.setAttribute(blockReader, Context.OperatorContext.METRICS_AGGREGATOR, blockReaderMetrics);
+    dag.setAttribute(blockReader, Context.OperatorContext.COUNTERS_AGGREGATOR, new BasicCounters.LongAggregator<MutableLong>());
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java]: Can you use METRICS_AGGREGATOR here? COUNTERS_AGGREGATOR is marked deprecated.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+
+/**
+ * HDFSFileSplitter extends {@link FileSplitterInput} to,
+ * 1. Add relative path to file metadata.
+ * 2. Ignore HDFS temp files (files with extensions _COPYING_).
+ * 3. Set sequencial read option on readers.
+ */
+public class HDFSFileSplitter extends FileSplitterInput
+{
+  private boolean sequencialFileRead;
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:37]: Does it make sense to have this property in FileSplitterInput operator itself if it is not specific to HDFS?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 13 above)
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: Can the relativePath property be added to super class? It seems generic enough.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+
+  /**
+   * HDFSScanner extends {@link TimeBasedDirectoryScanner} to ignore HDFS temporary files
+   * and files containing unsupported characters. 
+   */
+  public static class HDFSScanner extends TimeBasedDirectoryScanner
+  {
+    protected static final String HDFS_TEMP_FILE = ".*._COPYING_";
+    protected static final String UNSUPPORTED_CHARACTOR = ":";
+    private transient Pattern ignoreRegex;
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:72]: One way to handle this could be to expose ignoreRegex as a property for module, with default being _COPYING_. Could be useful ignore other file extensions as well. It can be also be part of superclass if it applies to other type of file readers.

File: library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+   * @param uri
+   * @return String with scheme part as lower case
+   */
+  private String convertSchemeToLowerCase(String uri)
+  {
+    if (uri == null) {
+      return null;
+    }
+    StringBuilder uriList = new StringBuilder();
+    for (String f : Splitter.on(",").omitEmptyStrings().split(uri)) {
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: Is it safe to assume the name of directory in a url cannot contain ',' ? Can list<uri> be a parameter instead?

File: library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+      String scheme = URI.create(f).getScheme();
+      if (scheme != null) {
+        uriList.append(f.replaceFirst(scheme, scheme.toLowerCase()));
```
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: Just a thought: Instead of converting scheme to lower case, can we throw an exception to indicate user to give correct uri?

Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: That's what happens by default we don't do anything. But there was a jira suggesting, we should not error out for such problem. we should be able to accept scheme irrespective of it's case.

File: library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java (hunk 23 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: We expect uses to configure input file(s)/directories separated by ",". Should we parse it in list during set method? Does it have any advantage?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 21 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:37]: It is not HDFS specific, but FileSplitterInput is too generic. If we decide this is good property for all purpose, I can move it to parent class.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 22 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:72]: We already expose a property called "filePatternRegularExp" for patterns to accept. Having one more property for patterns to reject, could be confusing so didn't expose it.
Is provided reason good enough not to expose it as property or I should expose it?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 13 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: Yes, definitely, I was just concerned if that would be accepted in generic class or not.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 14 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java]: By editor. In previous PR I had got a comment that better to use auto generated code for such methods.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 19 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java]: okay.

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 20 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java]: will update.

File: library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java (hunk 23 above)
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: A corner case, but what about the case when input directory is as below: 
hdfs://node18.morado.com:8020/user/isha/trial/comma,name

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java (hunk 22 above)
Comment (ishark) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:72]: I think ignoreFileRegex could be useful for non-hdfs file inputs as well.

Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSFileSplitter.java:72]: Okay I will expose this as property.

File: library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java (hunk 23 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/block/BlockReader.java]: good catch, how to we normally accept the multiple input then? Provided our primary source of properties is file based. Trying to figure out what else can be used as separator?

File: library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java (hunk 6 above)
Comment (DT-Priyanka) [inline library/src/main/java/com/datatorrent/lib/io/fs/HDFSInputModule.java:56]: I have created a Malhar jira for same: APEXMALHAR-2016, we would do it later and not part of this review request.
